proj/hps_accel/gateware/filter_store.py

```diff
@@ -80,21 +75,34 @@ def handle_writing(self, m, memory, num_words, index):
         with m.If(self.data_input.is_transferring()):
             m.d.sync += index.eq(index + 1)
             with m.If(index == num_words - 1):
-                m.d.sync += index.eq(0)
+                m.d.sync += index.eq(num_words - 4)
                 m.next = "READING"
 
-    def handle_reading(self, m, memory, num_words, index):
+    def handle_reading(self, m, memory, num_words, index, reset):
         """Handle stepping through memory on read."""
-        m.d.comb += [
-            memory.read_addr.eq(index[2:]),
-            self.data_output.payload.eq(memory.read_data),
-        ]
-        with m.If(self.next):
-            m.d.sync += self.data_output.valid.eq(1)
-        with m.If(~self.next & self.data_output.is_transferring()):
-            m.d.sync += self.data_output.valid.eq(0)
-        with m.If(self.data_output.is_transferring()):
-            m.d.sync += [index.eq(Mux(index == num_words - 4, 0, index + 4))]
+        m.d.comb += memory.read_addr.eq(index[2:])
+
+        read_port_valid = Signal()
+        read_started = Signal()
+
+        with m.If(~self.data_output.valid | self.data_output.is_transferring()):
+            m.d.sync += self.data_output.payload.eq(memory.read_data)
+            m.d.sync += self.data_output.valid.eq(read_port_valid)
+            m.d.sync += read_port_valid.eq(0)
+
+        with m.If(~read_port_valid & ~read_started):
```

Comment:
I think this is causing there to be 2 cycles between reads? Is that correct?

Reply:
Yes, this is the block which I spent too long staring at and rewriting. I'm still 90% sure that I should be able to trivially reduce this from 2 empty cycles to a single cycle, but I will need to spend more time staring at it so I'll leave it as is in this PR and send a follow-up later.